File: src/geeks_for_geeks/ds/union_find/UnionFind.java
package geeks_for_geeks.ds.union_find;

/**
 * Created By: Prashant Chaubey
 * Created On: 09-10-2018 01:07
 **/
public class UnionFind {
    private int parent[];
    private int rank[];

    public UnionFind(int size) {
        parent = new int[size];
        rank = new int[size];
        for (int i = 0; i < size; i++) {
            parent[i] = i;
        }
    }

    public int find(int val) {
        if (parent[val] != val) {
            //path compression
            parent[val] = find(parent[val]);
        }
        return parent[val];
    }

    public void union(int val1, int val2) {
        int parent1 = find(val1);
        int parent2 = find(val2);
        if (parent1 == parent2) {
            return;
        }
        if (rank[parent1] < rank[parent2]) {
            parent[parent1] = parent2;
        } else if (rank[parent2] < rank[parent1]) {
            parent[parent2] = parent1;
        } else {
            parent[parent1] = parent2;
            rank[parent2]++;
        }
    }

}

File: src/geeks_for_geeks/ds/union_find/union_find.py
class UnionFind:
    def __init__(self, size):
        self.parent = [0] * size
        self.rank = [0] * size
        self.disconnected = size
        for i in range(size):
            self.parent[i] = i

    def find(self, element):
        if self.parent[element] != element:
            self.parent[element] = self.find(self.parent[element])
        return self.parent[element]

    def same_set(self, first, second):
        parent1 = self.find(first)
        parent2 = self.find(second)
        if parent1 == parent2:
            return True
        return False

    def union(self, first, second):
        parent1 = self.find(first)
        parent2 = self.find(second)
        if parent1 == parent2:
            return
        if self.rank[parent1] < self.rank[parent2]:
            self.parent[parent1] = parent2
        elif self.rank[parent2] < self.rank[parent1]:
            self.parent[parent2] = parent1
        else:
            self.parent[parent2] = parent1
            self.rank[parent1] += 1
        self.disconnected -= 1
